Modules/_csv.c

```
@@ -238,21 +238,29 @@ _set_char(const char *name, Py_UCS4 *target, PyObject *src, Py_UCS4 dflt)
         if (src != Py_None) {
             Py_ssize_t len;
             if (!PyUnicode_Check(src)) {
-                PyErr_Format(PyExc_TypeError,
-                    "\"%s\" must be string, not %.200s", name,
-                    Py_TYPE(src)->tp_name);
+                if (strcmp(name, "delimiter") == 0) {
```

If we tweak this error message, we should also raise the same error if `delimiter=None`.

It may be simpler to introduce different function `_set_char_or_none()`.